src/filesystem/api.h

```diff
@@ -145,11 +145,15 @@ Status ReadTextFile(const std::string& path, std::string* contents);
 
 /// Create an object representing a local copy of a path.
 /// \param path The path of the directory or file.
+/// \param fetch_subdir If specified, will only download provided
+/// sub directory, otherwise all subdirectories will be downloaded.
+/// Does not affect files individual files, located under `path`.
 /// \param localized Returns the LocalizedPath object
 /// representing the local copy of the path.
 /// \return Error status
 Status LocalizePath(
-    const std::string& path, std::shared_ptr<LocalizedPath>* localized);
+    const std::string& path, const std::string& fetch_subdir,
+    std::shared_ptr<LocalizedPath>* localized);
```

Comment:
I think the interface is insufficiently generic for functions at the level of `LocalizePath()`, because `LocalizePath()` should be available to any callers who want to localize any path on the cloud, but `fetch_subdir` imposes a limitation on the caller to either fetch all sub-directories within the path or only the sub-directory specified. The caller may not "choose" which sub-directories to fetch or not fetch. let alone choosing sub-directories within a sub-directory.

I think there are many different generic design choices. For example:
1. Use `bool recursive` instead of `string fetch_subdir`. The caller can choose not fetching any sub-directories with `recursive = false` or all sub-directories with `recursive = true`. Then, the problem can be accomplished by first fetching the given model path with `recursive = false` (i.e. `/repo/model`) and then the model path with version with `recursive = true` (i.e. `/repo/model/1`). However, this will require the `LocalizedPath` object to be able to join the model path and model version path localization objects into one localization object.
2. Allow the caller to specify a list of sub-paths to localize (or not localize). The "complicated" structure of sub-paths allows for fetching different combinations of sub-directories in a single call, which eliminates the need for joining model path and model version path localization objects, but the "complicated" structure is less intuitive for the caller to use and may not be easier to implement than the first design choice.

However, the interface accomplishes exactly what we want to solve at this time. I will defer to @GuanLuo and @nnshah1 to decide.

Reply:
> However, this will require the LocalizedPath object to be able to join the model path and model version path localization objects into one localization object.
Do we? If I understand correctly, versioned path is set here: https://github.com/triton-inference-server/core/blob/a23786c25b36f07d2cfd510821682c117b50a99e/src/backend_model.cc#L113

so instead of 
```
const auto version_path =
      JoinPath({localized_model_path, std::to_string(version)});
```
we would do something like:
```
 std::shared_ptr<LocalizedPath> version_path;
  RETURN_IF_ERROR(LocalizePath(path/to/version, &version_path));
```
or am I missing something?

Reply:
I am thinking something like:
```
std::string model_path = "/repo/model_dir";
std::string model_version_path = "/repo/model_dir/1";

std::shared_ptr<LocalizedPath> model_localized;
std::shared_ptr<LocalizedPath> model_version_localized;
LocalizePath(model_path, recursive=false, &model_localized);
LocalizePath(model_version_path, recursive=true, &model_version_localized);

model_localized->Include(model_version_localized);
```

The idea behind `LocalizedPath::Include()` is to convert from
```
model_tmp
|
`--- config.pbtxt
```
and
```
model_version_tmp
|
`--- model.py
```
to
```
model_tmp
|
`--- config.pbtxt
|
`--- 1
     |
     `--- model.py
```
on the local disk.

Reply:
I actually am missing something. Even though I like option 1 suggestion, localization in this case will happen to another directory. Meaning, at current stage every call to `LocalizePath` creates a new temporary directory. Thus if we separately will try to load model and its version, they will be copied in 2 different directories. So unless we want to change that, option 1 is not the best option.

Reply:
I think it would be ok to only implement the base case above, and throw an error if the caller is trying to join/include anything more complicated.

Reply:
Based on offline discussion, with @kthui, we can add one more parameter `location` to control where to localize provided path, i.e. something like
```
LocalizePath(model_version_path, recursive-true/false, location=<string>, &model_version_localized);
```
Then we can control where version directory will be localized , and if `location` is empty, we create a new, temporary directory.

This will also help with the follow up ticket.

Reply:
> Meaning, at current stage every call to LocalizePath creates a new temporary directory. Thus if we separately will try to load model and its version, they will be copied in 2 different directories. So unless we want to change that, option 1 is not the best option.

I am thinking on the same direction as `LocalizedPath::Include()` that if we want to have a generic way to localizing directories (with customization on fetching subset of contents / recurrsiveness), we want to make `LocalizedPath` a "filler" class: as on calling `fs::LocalizePath`, it returns an fully localized object (current) or empty object (with local root path assigned). Then if user want to customize the localization, they can call:
```
LocalizedPath::Include("sub/path/content", recursive);
```
Although I think it then the process is convoluted that there is a iterative calls of `GetDirectorySubdirs/Files` and `LocalizedPath::Include`

Reply:
Note that this is really an internal API (unless we want to expose file system API in the future) so we are free to modify the interface as frequent as we want and don't need to worry about being generic. If localizing model directory is somewhat specific, we may just introduce a `LocalizeModelDirectory` method to fit the specific need.

Because when you look at what to be localized for a model version, it is actually "everything in the directory except for the unwanted version subdirectories", that is tricky to express it with some generic language.